src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1414,53 +1397,34 @@ private AuthenticationResult acquireTokenAfterValidation(CallbackHandler callbac
-        Logger.v(TAG, "Checking refresh tokens");
-        final boolean useMultiResourceRefreshToken = cachedItem == null || cachedItem.getIsMultiResourceRefreshToken();
-        final RefreshItem refreshItem = getRefreshToken(request, useMultiResourceRefreshToken);
+        
+        // If PromptBehavior is not "always" or "refresh_session", try silent flow first.
         AuthenticationResult authResult = null;
-        if (!promptUser(request.getPrompt()) && refreshItem != null
-                && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) {
-            Logger.v(TAG, "Refresh token is available and it will attempt to refresh token");
+        if (!promptUser(request.getPrompt())) {
             try {
-                authResult = getTokenWithRefreshTokenAndUpdateCache(request, refreshItem);
-            } catch (AuthenticationException authenticationException) {
-                ClientAnalytics.logEvent(
-                        InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED,
-                        new InstrumentationPropertiesBuilder(request, authenticationException).build());
+                final AcquireTokenSilentHandler acquireTokenSilentHandler = new AcquireTokenSilentHandler(mContext, 
+                        request, mTokenCacheAccessor);
+                authResult = acquireTokenSilentHandler.getAccessToken();
+            } catch (final AuthenticationException authenticationException) {
+                addLogEventForRequestFailureWithCause(InstrumentationIDs.REFRESH_TOKEN_REQUEST_FAILED, request, authenticationException);
                 callbackHandle.onError(authenticationException);
                 return null;
             }
             
-            if (authResult != null && !StringExtensions.IsNullOrBlank(authResult.getAccessToken())) {
+            if (isAccessTokenReturned(authResult)) {
                 callbackHandle.onSuccess(authResult);
                 ClientAnalytics.logEvent(
                         InstrumentationIDs.REFRESH_TOKEN_REQUEST_SUCCEEDED,
                         new InstrumentationPropertiesBuilder(request, authResult).build());
                 return authResult;
             }
-        } 
+        }
         
-        // refresh token does not exist or refresh token request failed to give back the token. 
+        // No token is returned.  
         // If it's non-silent request, will try to acquire token interactively. 
         // if it's silent request, will return the AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED.
-        if (refreshItem == null || authResult == null 
-                || (authResult != null && StringExtensions.IsNullOrBlank(authResult.getAccessToken()))) {
-            Logger.v(TAG, "Refresh token is not available or refresh token request failed to return token.");
+        if (!isAccessTokenReturned(authResult)) {
+            Logger.v(TAG, "No token returned for silent flow or promptbehavior is set to alwasys.");
```

Comment:
Done. 

---
In reply to: [63266473](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63266473) [](ancestors = 63266473)